test/integration/fake_upstream.cc

```diff
@@ -340,88 +349,46 @@ AssertionResult FakeConnectionBase::enableHalfClose(bool enable,
 }
 
 Http::RequestDecoder& FakeHttpConnection::newStream(Http::ResponseEncoder& encoder, bool) {
-  Thread::LockGuard lock(lock_);
+  absl::MutexLock lock(&lock_);
   new_streams_.emplace_back(new FakeStream(*this, encoder, time_system_));
-  connection_event_.notifyOne();
   return *new_streams_.back();
 }
 
-AssertionResult FakeConnectionBase::waitForDisconnect(bool ignore_spurious_events,
-                                                      milliseconds timeout) {
+AssertionResult FakeConnectionBase::waitForDisconnect(milliseconds timeout) {
   ENVOY_LOG(trace, "FakeConnectionBase waiting for disconnect");
-  auto end_time = time_system_.monotonicTime() + timeout;
-  Thread::LockGuard lock(lock_);
-  while (shared_connection_.connected()) {
-    if (time_system_.monotonicTime() >= end_time) {
-      return AssertionFailure() << "Timed out waiting for disconnect.";
-    }
-    Thread::CondVar::WaitStatus status = time_system_.waitFor(lock_, connection_event_, 5ms);
-    // The default behavior of waitForDisconnect is to assume the test cleanly
-    // calls waitForData, waitForNewStream, etc. to handle all events on the
-    // connection. If the caller explicitly notes that other events should be
-    // ignored, continue looping until a disconnect is detected. Otherwise fall
-    // through and hit the assert below.
-    if ((status == Thread::CondVar::WaitStatus::NoTimeout) && !ignore_spurious_events) {
-      break;
-    }
-  }
-
-  if (shared_connection_.connected()) {
-    return AssertionFailure() << "Expected disconnect, but got a different event.";
+  absl::MutexLock lock(&lock_);
+  const auto reached = [this]() {
+    lock_.AssertReaderHeld();
+    return !shared_connection_.connectedLockHeld();
+  };
+
+  if (!time_system_.waitFor(lock_, absl::Condition(&reached), timeout)) {
```

Comment:
did this ability to use a real lambda just get added?

Reply:
Not sure when it got added.